core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -919,15 +919,24 @@ class GroupCoordinator(val brokerId: Int,
 
   def onCompleteJoin(group: GroupMetadata) {
     group.inLock {
-      // remove any members who haven't joined the group yet
-      group.notYetRejoinedMembers.foreach { failedMember =>
+      // remove dynamic members who haven't joined the group yet
+      group.notYetRejoinedMembers.filter(!_.isStaticMember) foreach { failedMember =>
         removeHeartbeatForLeavingMember(group, failedMember)
         group.remove(failedMember.memberId)
         group.removeStaticMember(failedMember.groupInstanceId)
         // TODO: cut the socket connection to the client
       }
 
-      if (!group.is(Dead)) {
+      val joinedLeader = group.maybeElectNewJoinedLeader()
+      if (joinedLeader.isEmpty && group.allMembers.nonEmpty) {
+        // If all members are not rejoining, we will postpone the completion
+        // of rebalance preparing stage, and send out another dummy delayed operation
+        // until session timeout removes all the non-responsive members.
+        joinPurgatory.tryCompleteElseWatch(
+          new DelayedJoin(this, group, group.rebalanceTimeoutMs),
+          Seq(group.allMembers.headOption)
```

Comment:
Not sure I get this. Why are we using the memberId as the purgatory key?

Reply:
I think we could use a dummy key here.

Reply:
In fact, I do realize the purpose here. If this is last delayed join expiration, we will not trigger `onJoinComplete` again only depending on heartbeat timeout. So we have to constantly check for group emptiness